// telegram-api-bot/index.js
var fs = require('fs');
var https = require('https');

Object.extend = function(destination, source) {
    for (var property in source) {
        if (source.hasOwnProperty(property)) {
            destination[property] = source[property];
        }
    }
    return destination;
};

var telegram = function(token) {
  this.token = token;

  this.query = function(api, params, callback) {
    var params_s = "?";
    for(var k in params) {
      params_s+=encodeURIComponent(k)+"="+encodeURIComponent(params[k])+"&";
    }
    params_s=params_s.substring(0,params_s.length-1);
    var path = '/bot'+this.token+"/"+api+params_s;

    var options = {host: "api.telegram.org", path: path, port: 443};
    var data = '';
    https.get(options, function(res) {
      res.on('data', function(chunk){
        data+=chunk;
      });
      res.on('end', function(){
        if(callback!=undefined) {
          callback(JSON.parse(data.toString())["result"]);
        }
      });
    });
  };

  this.setWebhook = function(url, c) {
    this.query("setWebhook", {"url": url}, c);
  };

  this.getMe = function(c) {
    this.query("getMe", undefined, c);
  };

  this.sendMessage = function(chat_id, text, options, c) {
    var all_options = Object.extend({"chat_id":chat_id, "text":text}, options);
    this.query("sendMessage", all_options, c);
  };
  
  this.getFile = function(file_id, c) {
    this.query("getFile", {file_id: file_id}, c);
  };
  
  this.downloadFile = function(file_id, where, c) {
    var dest = where+"/"+file_id;
    var token = this.token;
    this.getFile(file_id, function(msg){
      var url = "https://api.telegram.org/file/bot"+token+"/"+msg.file_path;
      var file = fs.createWriteStream(where+"/"+file_id);
      var request = https.get(url, function(response) {
        response.pipe(file);
        file.on('finish', function() {
          file.close(function(){c(dest)});
        });
      }).on('error', function(err) {
        fs.unlink(dest); 
        if (c) c(err.message);
      });
    });
  };
};

module.exports = telegram;
